### PR TITLE
fix(security): harden fee payer protection against ATA rent drain and net-zero payment exploits

### PR DESCRIPTION
**Pull request by dev-jodee**
## Summary
- Block fee-payer-funded ATA `CreateIdempotent` instructions when `allow_create_account = false`, preventing post-sign ATA closure rent drains
- Conservatively include ATA rent in fee payer outflow even when simulation shows no inner `SystemCreateAccount` (stale-state protection)
- Enforce net payment delta instead of gross inflow credit — prevents attackers from looping tokens through the payment destination with zero net payment while draining SOL
- Extend net payment netting across bundle transactions to prevent cross-tx refund loops

## Test Plan
- `just check` — fmt, clippy, TS type-check all pass
- `just unit-test` — 660/660 pass
- `just integration-test` — all phases pass (regular, auth, payment_address, multi_signer, lighthouse, free_signing, fee_payer_policy, external, TS SDK)
- New unit tests: ATA parsing, policy rejection, conservative outflow accounting, no-double-count, `PaymentLamportTotals` netting, bundle cross-tx refund loop detection
- New integration tests: within-transaction net-zero token loop rejection, cross-bundle net-zero token loop rejection, SOL drain via net-zero loop rejection

## Breaking Changes (if applicable)
- None

---

<a href="https://app.devin.ai/review/solana-foundation/kora/pull/428" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

## 📊 Unit Test Coverage
Coverage

**Unit Test Coverage: 85.6%**

[View Detailed Coverage Report](https://github.com/solana-foundation/kora/actions/runs/24403333001)

### CODE REVIEW

File: crates/lib/src/fee/fee.rs [hunk 1]
```diff
@@ -389,6 +392,131 @@ impl FeeConfigUtil {
...
     }
 
+    async fn estimate_ata_account_len(
+        ata_creation: &AtaCreationInstructionInfo,
+        rpc_client: &RpcClient,
+        config: &Config,
+        token2022_account_len_cache: &mut HashMap<Pubkey, usize>,
+    ) -> Result<usize, KoraError> {
+        if ata_creation.token_program == spl_token_interface::id() {
+            return Ok(spl_token_interface::state::Account::LEN);
```
Comment (amilz) [inline crates/lib/src/fee/fee.rs:403]: do you want to add a `spl_token_2022_interface` import statement?

File: crates/lib/src/fee/fee.rs [hunk 2]
```diff
@@ -389,6 +392,131 @@ impl FeeConfigUtil {
...
+                spl_token_interface::state::Account::LEN
+            };
+
+            token2022_account_len_cache.insert(ata_creation.mint, account_len);
+            return Ok(account_len);
+        }
+
+        // Unsupported token programs should fail on-chain, but use conservative SPL account length
+        // for safety if they are present in a transaction we are estimating.
+        Ok(spl_token_interface::state::Account::LEN)
```
Comment (amilz) [inline crates/lib/src/fee/fee.rs]: this is not actually conservative...it's the lowest option. IMO unless we have backsups like this for other token operations in kora we should probably just throw?

File: crates/lib/src/fee/fee.rs [hunk 3]
```diff
@@ -389,6 +392,131 @@ impl FeeConfigUtil {
...
+            if let Some(account_len) = token2022_account_len_cache.get(&ata_creation.mint) {
+                return Ok(*account_len);
+            }
+
+            let mint_state = TokenUtil::get_mint(config, rpc_client, &ata_creation.mint).await?;
+            let account_len = if let Some(token2022_mint) =
+                mint_state.as_any().downcast_ref::<Token2022Mint>()
+            {
+                let required_account_extensions =
+                    spl_token_2022_interface::extension::ExtensionType::get_required_init_account_extensions(
```
Comment (amilz) [inline crates/lib/src/fee/fee.rs]: i think ata always using immutable owner. 

https://github.com/solana-program/associated-token-account/blob/7af39d84438199e7e488adc379baa8ee0b8085c0/program/src/processor.rs#L120-L124

File: crates/lib/src/token/token.rs [hunk 4]
```diff
@@ -57,6 +57,45 @@ impl TokenType {
...
+    pub(crate) outflow: u64,
+}
+
+impl PaymentLamportTotals {
+    pub(crate) fn net_payment(self) -> Option<u64> {
+        let net_payment_lamports = self.inflow.saturating_sub(self.outflow);
+        if net_payment_lamports > 0 {
+            Some(net_payment_lamports)
+        } else {
+            None
```
Comment (amilz) [inline crates/lib/src/token/token.rs]: Any reason not to just return `u64` and make this 0 (instead of `Option`)?

File: crates/lib/src/token/token.rs [hunk 5]
```diff
@@ -261,29 +310,68 @@ impl TokenUtil {
...
         }
         None
     }
 
+    /// Parse ATA Create/CreateIdempotent instructions.
+    /// Returns None for non-ATA instructions or unsupported ATA variants.
+    pub fn parse_ata_creation_instruction(
+        instruction: &Instruction,
+    ) -> Option<AtaCreationInstructionInfo> {
+        let ata_program_id = spl_associated_token_account_interface::program::id();
```
Comment (amilz) [inline crates/lib/src/token/token.rs]: do you want to import this stuff at top of file?

File: crates/lib/src/constant.rs [hunk 6]
```diff
@@ -221,9 +222,11 @@ pub mod instruction_indexes {
 
     // ATA instruction indexes
```
Comment (amilz) [inline crates/lib/src/constant.rs]: IMO since you have everything else already, I'd just include system program (index 4 here) and a link to the source in comments: https://github.com/solana-program/associated-token-account/blob/7af39d84438199e7e488adc379baa8ee0b8085c0/interface/src/instruction.rs#L19

File: crates/lib/src/token/token.rs [hunk 7]
```diff
@@ -715,80 +830,76 @@ impl TokenUtil {
...
+                    })?;
 
-                // Skip unsupported tokens
                 if !config.validation.supports_token(&token_mint.to_string()) {
                     log::warn!("Ignoring payment with unsupported token mint: {}", token_mint,);
                     continue;
                 }
 
-                let effective_amount = if *is_2022 {
+                if *is_2022 && is_inflow && !is_outflow {
```
Comment (amilz) [inline crates/lib/src/token/token.rs]: Would we not want to validate regardless of in/outflow status?

File: crates/lib/src/token/token.rs [hunk 8]
```diff
@@ -57,6 +57,45 @@ impl TokenType {
...
 }
 
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub(crate) struct PaymentLamportTotals {
+    pub(crate) inflow: u64,
+    pub(crate) outflow: u64,
+}
+
+impl PaymentLamportTotals {
+    pub(crate) fn net_payment(self) -> Option<u64> {
```
Comment (amilz) [inline crates/lib/src/token/token.rs]: wdyt about a warning log for helping folks recognize undesirable behavior

```rust
if self.outflow > self.inflow {
          log::warn!(
              "Net-negative payment detected (possible drain attempt): inflow={}, outflow={}",
              self.inflow,
              self.outflow
          );
      }
```

File: crates/lib/src/token/token.rs [hunk 9]
```diff
@@ -261,29 +310,68 @@ impl TokenUtil {
...
+    ) -> Option<AtaCreationInstructionInfo> {
+        let ata_program_id = spl_associated_token_account_interface::program::id();
+        if instruction.program_id != ata_program_id
+            || instruction.accounts.len()
+                < constant::instruction_indexes::ata_instruction_indexes::MIN_ACCOUNTS
+        {
+            return None;
+        }
+
+        let discriminator = instruction.data.first().copied().unwrap_or(0);
```
Comment (amilz) [inline crates/lib/src/token/token.rs]: should this throw rather than defaulting to 0?

File: crates/lib/src/token/token.rs [hunk 10]
```diff
@@ -1702,4 +1865,84 @@ mod tests_token {
         let result = TokenUtil::find_ata_creation_for_destination(&instructions, &target_address);
         assert!(result.is_none());
     }
+
+    #[test]
+    fn test_parse_ata_creation_instruction_idempotent() {
```
Comment (amilz) [inline crates/lib/src/token/token.rs:1943]: Should add non-idempotentn test as well?

File: crates/lib/src/token/token.rs [hunk 11]
```diff
@@ -715,80 +830,76 @@ impl TokenUtil {
...
-                // Skip if destination isn't our expected payment address
-                if destination_owner != *expected_destination_owner {
+                let is_inflow = destination_account_info
+                    .as_ref()
+                    .map(|(owner, _, _)| owner == expected_destination_owner)
+                    .unwrap_or(false);
+                let is_outflow = source_account_info
+                    .as_ref()
+                    .map(|(owner, _, _)| owner == expected_destination_owner)
+                    .unwrap_or(false);
```
Comment (amilz) [inline crates/lib/src/token/token.rs:862]: might be worth adding a test case where our expected destination matches both

File: crates/lib/src/fee/fee.rs [hunk 12]
```diff
@@ -389,6 +392,131 @@ impl FeeConfigUtil {
...
+                .filter_map(|instruction| match instruction {
+                    ParsedSystemInstructionData::SystemCreateAccount {
+                        payer, new_account, ..
+                    } if *payer == *fee_payer_pubkey => Some(*new_account),
+                    _ => None,
+                })
+                .collect();
+        let mut rent_cache_by_account_len: HashMap<usize, u64> = HashMap::new();
+        let mut token2022_account_len_cache: HashMap<Pubkey, usize> = HashMap::new();
+        let mut seen_ata_addresses = HashSet::new();
```
Comment (amilz) [inline crates/lib/src/fee/fee.rs:474]: Should we add some test coverages for multiple ata ix in single txn? Edge cases like multiple of the same?

File: tests/adversarial/fee_payer_exploitation.rs [hunk 13]
```diff
@@ -43,6 +48,77 @@ async fn test_fee_payer_as_sol_transfer_source() {
     }
 }
```
Comment (amilz) [inline tests/adversarial/fee_payer_exploitation.rs:51]: wdyt about adding t22 coverage here as well?
